### PR TITLE
Don't do state cleanup if training not converged and return trainer's state

### PR DESCRIPTION
**Pull request by artemsok**
By default the ```training_state``` folder is deleted even if the training did not converge, i.e. the number of checkpoints did not hit ```--max-num-not-improved```, but the training stopped because of reaching ```--max-samples```, ```--max-num-epochs```, or ```--max-updates```. This changes the behaviour to keeping the folder to allow later training continuation and not enforcing the above parameters to match with the old run. Additionally a full training state is returned (with metrics, number of epochs, convergence status etc.), allowing its processing in a calling code.

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sockeye/constants.py [hunk 1]
```diff
@@ -238,7 +238,8 @@
 # Arguments that may differ and still resume training
 ARGS_MAY_DIFFER = ["overwrite_output", "use-tensorboard", "quiet",
                    "align_plot_prefix", "sure_align_threshold",
-                   "keep_last_params"]
+                   "keep_last_params", "seed", "max-updates",
+                   "max-num-epochs", "max-samples"]
```
Comment (fhieber) [inline sockeye/constants.py]: How about `--min-num-*`? In general, would it make sense to make a pass over current arguments to see whether we should add more to this list? Its a suboptimal solution as it becomes outdated quickly, but if you are at it anyway right now :)

File: sockeye/training.py [hunk 2]
```diff
@@ -761,6 +767,9 @@ def _cleanup(self, lr_decay_opt_states_reset: str, process_manager: Optional['De
                 self.tflogger.log_metrics(decoder_metrics, decoded_checkpoint)
             utils.write_metrics_file(self.state.metrics, self.metrics_fname)
 
+        if keep_training_state:
```
Comment (fhieber) [inline sockeye/training.py]: matter of taste, but I would prefer to move the following lines under an `if not keep_training_state` instead of the early return.

File: sockeye/training.py [hunk 3]
```diff
@@ -523,6 +523,8 @@ def fit(self,
         speedometer = Speedometer(frequency=C.MEASURE_SPEED_EVERY, auto_reset=False)
         tic = time.time()
 
+        converged = False
```
Comment (fhieber) [inline sockeye/training.py]: not really important, but would it make sense to move this boolean into `TrainState` just for clarity?
In that case, this function could even return the TrainState object, which holds all kinds of information from training (best metric, updates, epochs etc.)

File: sockeye/train.py [hunk 4]
```diff
@@ -863,25 +863,25 @@ def train(args: argparse.Namespace):
...
-                    min_epochs=min_epochs,
-                    max_epochs=max_epochs,
-                    lr_decay_param_reset=args.learning_rate_decay_param_reset,
-                    lr_decay_opt_states_reset=args.learning_rate_decay_optimizer_states_reset,
-                    decoder=create_checkpoint_decoder(args, exit_stack, context),
-                    mxmonitor_pattern=args.monitor_pattern,
-                    mxmonitor_stat_func=args.monitor_stat_func,
-                    allow_missing_parameters=args.allow_missing_params or model_config.lhuc,
-                    existing_parameters=args.params)
+        return trainer.fit(train_iter=train_iter,
```
Comment (fhieber) [inline sockeye/train.py]: [nit] I'd prefer a more explicit style here, i.e.
```python
best_metric, converged = trainer.fit(...)
return best_metric, converged
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -10,6 +10,11 @@ Note that Sockeye has checks in place to not translate with an old model that wa
 
 Each version section may have have subsections for: _Added_, _Changed_, _Removed_, _Deprecated_, and _Fixed_.
 
+## [1.18.59]
+- A flag is returned alongside the best metric indicating if training converged.
```
Comment (fhieber) [inline CHANGELOG.md]: I'd put this under an `### Added section`.

Comment (artemsok) [reply CHANGELOG.md]: Oh, it was eaten up by a merge, returned.

File: sockeye/train.py (hunk 4 above)
Comment (artemsok) [reply sockeye/train.py]: Done

File: sockeye/training.py (hunk 3 above)
Comment (artemsok) [reply sockeye/training.py]: Makes sense, actually this may simplify calling training from some code.

File: sockeye/training.py (hunk 2 above)
Comment (artemsok) [reply sockeye/training.py]: fixed

File: sockeye/constants.py (hunk 1 above)
Comment (artemsok) [reply sockeye/constants.py]: Agree, ```--min-*``` shouldn't matter either, added them.

File: CHANGELOG.md [hunk 6]
```diff
@@ -10,6 +10,12 @@ Note that Sockeye has checks in place to not translate with an old model that wa
 
 Each version section may have have subsections for: _Added_, _Changed_, _Removed_, _Deprecated_, and _Fixed_.
 
+## [1.18.59]
+### Added
+- Full training state is now returned from EarlyStoppingTrainer's fit().
+### Changed
+- Training state cleanup will not be performed for training runs that did not converge.
```
Comment (tdomhan) [inline CHANGELOG.md]: Maybe add a 'yet': 'that did not converge **yet**'

File: sockeye/training.py [hunk 7]
```diff
@@ -455,7 +456,7 @@ def fit(self,
             mxmonitor_pattern: Optional[str] = None,
             mxmonitor_stat_func: Optional[str] = None,
             allow_missing_parameters: bool = False,
-            existing_parameters: Optional[str] = None):
+            existing_parameters: Optional[str] = None) -> 'TrainState':
```
Comment (tdomhan) [inline sockeye/training.py]: No need to forward reference as the TrainState is defined above, no?
```suggestion
            existing_parameters: Optional[str] = None) -> TrainState:
```

File: CHANGELOG.md (hunk 6 above)
Comment (artemsok) [reply CHANGELOG.md]: fixed

File: sockeye/training.py (hunk 7 above)
Comment (artemsok) [reply sockeye/training.py]: thanks, indeed
